### PR TITLE
lib.customisation: uncurry makeScopeWithSplicing

### PR DESCRIPTION
**Pull request by ghost**
###### Description of changes

Deeply-curried functions are pretty error-prone in untyped languages like Nix.  This is a particularly bad case because `top-level/splice.nix` *also* declares a makeScopeWithSplicing, but it takes *two fewer arguments*.

Let's switch to attrset-passing form, to provide some minimal level of sanity-checking.

###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: lib/customisation.nix
Comment (Artturin) [inline lib/customisation.nix:1]:
```diff
diff --git a/lib/customisation.nix b/lib/customisation.nix
index a46913dc5bde..4dfe75023b96 100644
--- a/lib/customisation.nix
+++ b/lib/customisation.nix
@@ -279,7 +279,7 @@ rec {
 
   /* Like the above, but aims to support cross compilation. It's still ugly, but
      hopefully it helps a little bit. */
-  makeScopeWithSplicing = { splicePackages, newScope }: { otherSplices, keep, extra, f }:
+  makeScopeWithSplicing = { splicePackages, newScope, otherSplices, keep, extra, f }:
     let
       spliced0 = splicePackages {
         pkgsBuildBuild = otherSplices.selfBuildBuild;
@@ -296,8 +296,7 @@ rec {
         # N.B. the other stages of the package set spliced in are *not*
         # overridden.
         overrideScope = g: (makeScopeWithSplicing
-          { inherit splicePackages newScope; }
-          { inherit otherSplices keep extra;
+          { inherit splicePackages newScope otherSplices keep extra;
             f = lib.fixedPoints.extends g f;
           });
         packages = f;
diff --git a/pkgs/top-level/splice.nix b/pkgs/top-level/splice.nix
index 57cedb12a2f5..01407c107a16 100644
--- a/pkgs/top-level/splice.nix
+++ b/pkgs/top-level/splice.nix
@@ -144,7 +144,7 @@ in
   newScope = extra: lib.callPackageWith (splicedPackagesWithXorg // extra);
 
   # prefill 2 fields of the function for convenience
-  makeScopeWithSplicing = lib.makeScopeWithSplicing { inherit splicePackages; inherit (pkgs) newScope; };
+  makeScopeWithSplicing = a: lib.makeScopeWithSplicing ({ inherit splicePackages; inherit (pkgs) newScope; } // a);
 
   # generate 'otherSplices' for 'makeScopeWithSplicing'
   generateSplicesForMkScope = attr:
```

Alternative which would not require 2 input sets, which looks nicer if lib.makeScopeWithSplicing is used directly, IMO.

Comment (Artturin) [reply lib/customisation.nix:1]: functionArgs won't work on pkgs.makeScopeWithSplicing though

**Review by Artturin — APPROVED** [review]

**Review by ghost — COMMENTED** [review]
I guess I can't "approve" my own PR, but I approve of the commit @Artturin pushed onto it.

**Comment by infinisil** [comment]
This is not okay, this is a breaking change to `lib` which breaks all third-party uses of it

**Comment by amjoseph-nixpkgs** [comment]
> This is not okay, this is a breaking change to `lib` which breaks all third-party uses of it

You're right.  For some reason I thought `makeScopeWithSplicing` was added recently, but I was wrong about that.

Will revert and reintroduce with a new name for the uncurried version.

**Comment by amjoseph-nixpkgs** [comment]
I have merged https://github.com/NixOS/nixpkgs/pull/245955

**Comment by infinisil** [comment]
Thanks!

**Comment by amjoseph-nixpkgs** [comment]
Reimplemented with a different name and a backward-compatible shim: https://github.com/NixOS/nixpkgs/pull/245957

**Pull request merged**